### DISCUSSION TITLE
This time a simple question - does peek update statistics for cache misses / hits or do only get & getAll do this?

### DISCUSSION THREAD

**Discussion by javafanboy** [oracle/coherence#90] [General]
I am deep diving into some performance results we are seeing.  In some situations we do "peek" towards our front cache (in a near cache setup) to determine how to proceed in our execution (favoring in some cases to do bulk requests of items not in the front cache over retrieving them one by one) .

I would naively have assumed that only get & getAll requests update the hit/miss statistics but one of my coworkers asked me if peek could also play in here (i.e. when peek return null is this recorded as a miss and when it returns a value as a hit?) so to be 100% sure how to interpret the results we see I wanted to confirm this....

I suppose it would be good to make this clear in the documentation of the peek method also it, at least to us, was not obvious....

**Comment by javafanboy** [comment]
To be clear I think it is unfortunate if peek does not update these counters for the front cache as it, in my view, is a fully valid alternative to using get. If using peek a miss (returning null) should of course not update any statistics on the back cache though...

**Reply by aseovic** [reply]
Sometimes the best thing to do is look at the actual implementation ;-)

https://github.com/oracle/coherence/blob/master/prj/coherence-core/src/main/java/com/tangosol/net/cache/LocalCache.java#L463

But to answer your question, no, `peek` doesn't update the statistics, as the comment in it explicitly states:
```java
    public Object peek(Object oKey)
        {
        // avoid super.get() because it affects statistics
        OldCache.Entry entry = (OldCache.Entry) getEntryInternal(oKey);
        return entry == null ? null : entry.getValue();
        }
```
The stats are updated within `getEntry` method, based on the result of the `getEntryInternal` call, the same call `peek` uses:
```java
    public SafeHashMap.Entry getEntry(Object oKey)
        {
        // check if the cache needs flushing
        tryEvict();

        Entry entry = (Entry) getEntryInternal(oKey);
        if (entry == null)
            {
            m_stats.registerMiss();
            }
        else
            {
            m_stats.registerHit();
            entry.touch();
            }

        return entry;
        }
```

**Reply by aseovic** [reply]
Of course, nothing prevents you from extending `LocalCache` and overriding `peek` to update statistics accordingly, if that's what you really want.

You'd just have to switch from using the standard `local-scheme` to your custom `class-scheme` for your front cache.

**Reply by aseovic** [reply]
Finally, keep in mind that `peek` is `LocalCache`-specific operation that is not exposed via `CacheMap` interface, so it may not be available in other local cache/backing map implementations, such as `CaffeineCache` (and indeed, it isn't). 

Basically, it means you are tying your code to `LocalCache`.

**Comment by javafanboy** [comment]
Thanks for the info and good suggestion with the subclassing.
